addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/cosem/CosemDate.java

```diff
@@ -0,0 +1,128 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.dsmr.internal.device.cosem;
+
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.eclipse.smarthome.core.library.types.DateTimeType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * CosemDate represents a datetime value and will try to autodetect the format
+ *
+ * @author M. Volaart
+ * @since 2.1.0
+ */
+public class CosemDate extends CosemValue<Date> {
+    // logger
+    private final Logger logger = LoggerFactory.getLogger(CosemDate.class);
+
+    /**
+     * This enum contains the known date formats for the DSMR-specification
+     */
+    private enum CosemDateFormat {
+        /*
+         * Ignore DST setting for general format. We use local time that is
+         * already DST
+         */
+        COSEM_DATE_GENERAL("(\\d{12})([S,W]?)", "yyMMddHHmmss"),
+        COSEM_DATE_DSMR_V2("(\\d{2}-\\d{2}-\\d{2} \\d{2}:\\d{2}:\\d{2})", "yy-MM-dd HH:mm:ss");
+
+        /* Cached compiled pattern */
+        private final Pattern pattern;
+
+        /* Cached java date formatter */
+        private final SimpleDateFormat formatter;
```

Comment:
Please note that `SimpleDateFormat` is NOT thread safe, Java-8's DateTimeFormatter is your friend in this case.